src/python/pants/backend/go/module.py

```diff
@@ -0,0 +1,203 @@
+    flattened_sources_digest = await Get(
+        Digest, RemovePrefix(sources.snapshot.digest, request.address.spec_path)
+    )
+    flattened_sources_snapshot = await Get(Snapshot, Digest, flattened_sources_digest)
+    if (
+        len(flattened_sources_snapshot.files) not in (1, 2)
+        or "go.mod" not in flattened_sources_snapshot.files
+    ):
+        raise ValueError(f"Incomplete go_module sources: files={flattened_sources_snapshot.files}")
+
+    analyze_script_digest = await Get(
+        Digest,
+        CreateDigest(
+            [
+                FileContent(
+                    "analyze.sh",
+                    textwrap.dedent(
+                        """\
+                export GOROOT="./go"
+                export GOPATH="$(/bin/pwd)/gopath"
+                export GOCACHE="$(/bin/pwd)/cache"
+                mkdir -p "$GOPATH" "$GOCACHE"
+                exec ./go/bin/go mod download -json all
+                """
+                    ).encode("utf-8"),
+                )
+            ]
+        ),
+    )
+
+    input_root_digest = await Get(
+        Digest,
+        MergeDigests([flattened_sources_digest, downloaded_goroot.digest, analyze_script_digest]),
+    )
+
+    process = Process(
+        argv=[bash.path, "./analyze.sh"],
+        input_digest=input_root_digest,
+        description="Resolve go_module metadata.",
+        output_files=["go.mod", "go.sum"],
+        level=LogLevel.DEBUG,
+    )
+
+    result = await Get(ProcessResult, Process, process)
+
+    # Parse the go.mod for the module path and minimum Go version.
+    module_path = None
+    minimum_go_version = None
+    digest_contents = await Get(DigestContents, Digest, flattened_sources_digest)
+    for entry in digest_contents:
+        if entry.path == "go.mod":
+            module_path, minimum_go_version = basic_parse_go_mod(entry.content)
+
+    if module_path is None:
+        raise ValueError("No `module` directive found in go.mod.")
+
+    return ResolvedGoModule(
+        import_path=module_path,
+        minimum_go_version=minimum_go_version,
+        modules=FrozenOrderedSet(parse_module_descriptors(result.stdout)),
+        digest=result.output_digest,
+    )
+
+
+# TODO: Add integration tests for the `go-resolve` goal once we figure out its final form. For now, it is a debug
+# tool to help update go.sum while developing the Go plugin and will probably change.
+class GoResolveSubsystem(GoalSubsystem):
+    name = "go-resolve"
+    help = "Resolve a Go module's go.mod and update go.sum accordingly."
+
+
+class GoResolveGoal(Goal):
+    subsystem_cls = GoResolveSubsystem
+
+
+@goal_rule
+async def run_go_resolve(
+    targets: UnexpandedTargets, console: Console, workspace: Workspace
+) -> GoResolveGoal:
+    for target in targets:
+        if target.has_field(GoModuleSources) and not target.address.is_file_target:
+            resolved_go_module = await Get(ResolvedGoModule, ResolveGoModuleRequest(target.address))
```

Comment:
Using `Get` in a loop misses out on a lot of performance. Better is to use `await MultiGet`. https://www.pantsbuild.org/docs/rules-api-tips#tip-use-multiget-for-increased-concurrency

Reply:
That is an optimization to do for cleanup phase. During this development phase, it is easier for me to reason about the code if I am not also trying to make stuff concurrent.